### PR TITLE
Update Docker Deployment

### PR DESCRIPTION
**Pull request by NOXCIS**
- Configs & Peers Persist Across Container Updates
- Containers Stop time Reduction via Docker Sigterm and existing wgd.sh stop function
- Added Environment file for passing Extra Docker Args
- Updated Docker Readme
- Fixed HealthCheck
- ARM 32 bit support.

### CODE REVIEW

**Comment by DaanSelen** [comment]
@donaldzou if you have the time, I can look at it next week?

**Comment by donaldzou** [comment]
Hi @DaanSelen, if you wish, do mind look over it since I'm no expert of Docker haha, but if you need to assist on the dashboard itself, I'm happy to help :)

**Comment by DaanSelen** [comment]
> Hi @DaanSelen, if you wish, do mind look over it since I'm no expert of Docker haha, but if you need to assist on the dashboard itself, I'm happy to help :)

I am a bit thorn between the two images, perhaps offer both Docker image as resources on the main page. But for you I would like your vision of what the Docker Image of WGDashboard should be, I made my image of course in my vision, and NOXCIS in his I assume.

**Comment by donaldzou** [comment]
> > Hi @DaanSelen, if you wish, do mind look over it since I'm no expert of Docker haha, but if you need to assist on the dashboard itself, I'm happy to help :)
> 
> I am a bit thorn between the two images, perhaps offer both Docker image as resources on the main page. But for you I would like your vision of what the Docker Image of WGDashboard should be, I made my image of course in my vision, and NOXCIS in his I assume.

Of course. I will spend sometime going through it let you know :)

**Comment by DaanSelen** [comment]
> > > Hi @DaanSelen, if you wish, do mind look over it since I'm no expert of Docker haha, but if you need to assist on the dashboard itself, I'm happy to help :)
> > 
> > 
> > I am a bit thorn between the two images, perhaps offer both Docker image as resources on the main page. But for you I would like your vision of what the Docker Image of WGDashboard should be, I made my image of course in my vision, and NOXCIS in his I assume.
> 
> Of course. I will spend sometime going through it let you know :)

Alternatively you can cherrypick features you like about each image and try to make a merger.

**Comment by NOXCIS** [comment]
@DaanSelen @donaldzou The Dashboard already has everthing need within it self to deploy on a sysytem. i just made a wrpper with docker that allows args to be passsed securely directly before wgdashboard.ini is created.  Leveraging heavily off of the exsisting wgd.sh script.

**Comment by donaldzou** [comment]
Hi @NOXCIS , after viewing carefully on the Docker changes, I'm thinking would it be possible to have less changes on `wgd.sh` and `dashboard.py`? I'm hoping the Docker implementation does not affect the source code and adding complexity to it. What do you think?

**Comment by NOXCIS** [comment]
@donaldzou Understandable, at the moment I'm not really sure of a way of securely passing environment variables to the wg-dashboard.ini other than an environment file. Because of the Python venv docker can't pass args from shell with the only other option being to bake the environment into the image. Which would require users to build the docker image on their own instead of pulling a prebuilt image. 

That's the catch 22 I tried to work around by adding dotenv to the dashboard to load the environment args from the .env created by wgd.sh with passed docker environment and default for non docker users. The venv is activated and dotenv loads the environment from .env . It's ignored after first install because wg-dashboard.ini would already exist. 

In short the environment loading can be removed and still maintain full functionality but at the cost of a predetermined environment. I.E users have to set everything in settings after startup. What do you think?

**Comment by donaldzou** [comment]
Honestly I think is better let user configure after startup.. just to keep everything simple ;)

**Comment by NOXCIS** [comment]
@donaldzou cool remove what you don't want and use the rest or do whatever. My primary specialty is security researcher so I will have to hard fork again. The docker image is being used as a compose again and I'm not one for circular arguments. Thanks for working on the dashboard.

**Comment by DaanSelen** [comment]
Hi Noxcis, I do not want to understand your contributions! You've actually provided a lot of information. And I do invite you create an image alongside the main one, instead of replacing the one. So people have a choice. I do value your opinion on subjects.

The main concern I noticed is that the Docker image was changing the code, what should be the other way around, in my opinion. If you make a max-secure image we can integrate it nicely.

The way everything went down is not optimal so I want to create a solution.

**Comment by NOXCIS** [comment]
@DaanSelen Not sure what you mean by changing the code. Either way the image can't be updated anymore  due to the config being generated upon building the image itself without deploying a container. 
Essentially the issue that cause the security vulnerability in the first place. 

I added environment loading so that variables can be set via a .env so I don't have to hard code into the image. This is what's being done over on [wg-easy](https://github.com/wg-easy/wg-easy) the most popular WireGuard dashboard on GitHub. The environment loading can be removed      And functionality can persist. 

However what I cannot deploy in my  environments is a patch work around the same hardcoded Variables. 

- hard set PostUP & PostDown rules that require a image rebuild or manually entering the container to modify.

- Any updates to the dashboard will require an image rebuild and compose up for the end user. Which will nuke the wg0.conf created during build. 

- I prefer WG-Dashboard over WG-easy because of the increased functionality and design. However the current docker deployment for WG Dashboard falls far behind all other WireGuard dashboards on GitHub without the compose defined environment.

- The WG-Dashboard docker deployment in its current state is  insufficient when compared to all other WireGuard UI's on GitHub. Take a look around at their Dockerfiles for inspiration.

- In short the docker deployment without proper environment setting as per the docker documentation defeats dockers purpose as a tool that allows easy deployment. 

- I will continue my work over on Wiregate. It's been fun working with the dashboard directly. Use whatever code you guys like and throw away what you don't :).

**Comment by DaanSelen** [comment]
> @DaanSelen Not sure what you mean by changing the code. Either way the image can't be updated anymore due to the config being generated upon building the image itself without deploying a container. Essentially the issue that cause the security vulnerability in the first place.
> 
> I added environment loading so that variables can be set via a .env so I don't have to hard code into the image. This is what's being done over on [wg-easy](https://github.com/wg-easy/wg-easy) the most popular WireGuard dashboard on GitHub. The environment loading can be removed And functionality can persist.
> 
> However what I cannot deploy in my environments is a patch work around the same hardcoded Variables.
> 
> * hard set PostUP & PostDown rules that require a image rebuild or manually entering the container to modify.
> * Any updates to the dashboard will require an image rebuild and compose up for the end user. Which will nuke the wg0.conf created during build.
> * I prefer WG-Dashboard over WG-easy because of the increased functionality and design. However the current docker deployment for WG Dashboard falls far behind all other WireGuard dashboards on GitHub without the compose defined environment.
> * The WG-Dashboard docker deployment in its current state is  insufficient when compared to all other WireGuard UI's on GitHub. Take a look around at their Dockerfiles for inspiration.
> * In short the docker deployment without proper environment setting as per the docker documentation defeats dockers purpose as a tool that allows easy deployment.
> * I will continue my work over on Wiregate. It's been fun working with the dashboard directly. Use whatever code you guys like and throw away what you don't :).

Hello @NOXCIS,
I think we are confusing general application problems with container problems.

- Hard set PostUP & PostDown rules; Is this an issue of the container or of WGDashboard application itself, is it possible to edit configurations once they were already existing? I will test this and edit the reply.
- Why will it nuke the config created on the build? The entrypoint stats the file and does nothing when it detects the wg0.conf. Line 33 - 47 `entrypoint.sh`
- How does it fall behind? It's a fully working Docker Container?
- Please bring up actual examples so we can work with that.
- You can run it completely without any environment variables, and it will use defaults that are defined?
- You are always welcome to work on the main project.

Kind regards, Daan Selen.

**Comment by NOXCIS** [comment]
@DaanSelen WireGuard doesnt have an official user interface, hence we have to rely on opensource. Take your time and look at the dockerfile and compose of each of the mainstream open source options.

1. [wg-easy](https://github.com/wg-easy/wg-easy.git) 14.5K stars
2. [wireguard](https://github.com/ngoduykhanh/wireguard-ui) 4K stars
3. [wag](https://github.com/NHAS/wag) 505 stars
4. [wireadmin](https://github.com/wireadmin/wireadmin.git) 274 stars
5. [wg-gen-web](https://github.com/vx3r/wg-gen-web) 1.6K stars


Out of all of the above repos no dev/ contributor has hard coded the wireguard config geration into the docker image. Fun fact they never did anything to result in a mass spead of the same config to all endusers either. That would be a terrible vulnerabilty to have.  Someone would have to be a self labeled ignorant dev to do such thing. Anyways cheers, ive done my part by giving back to opensorce for the next generation.

**Comment by DaanSelen** [comment]
> @DaanSelen WireGuard doesnt have an official user interface, hence we have to rely on opensource. Take your time and look at the dockerfile and compose of each of the mainstream open source options.
> 
> 1. [wg-easy](https://github.com/wg-easy/wg-easy.git) 14.5K stars
> 2. [wireguard](https://github.com/ngoduykhanh/wireguard-ui) 4K stars
> 3. [wag](https://github.com/NHAS/wag) 505 stars
> 4. [wireadmin](https://github.com/wireadmin/wireadmin.git) 274 stars
> 5. [wg-gen-web](https://github.com/vx3r/wg-gen-web) 1.6K stars
> 
> Out of all of the above repos no dev/ contributor has hard coded the wireguard config geration into the docker image. Fun fact they never did anything to result in a mass spead of the same config to all endusers either. That would be a terrible vulnerabilty to have. Someone would have to be a self labeled ignorant dev to do such thing. Anyways cheers, ive done my part by giving back to opensorce for the next generation.

Calling names does not further the project. Give arguments for your choices, not just "because it is", please.
I gave you questions and you just give these links without any arguments?

The private key was in one of the first iterations of the container and was fixed, just how it should be? There was a notice helping people migrate? I understand the frustration but what was provided in return, was a non-functional product.

Once WGDashboard gets the ability to generate its configuration files itself, then the hard-core part is not needed anymore. But right now its not there yet, as is normal with software development. Overall I see not many arguments for the choices, which is unfortunate.

**Comment by NOXCIS** [comment]
"Hi Noxcis, I do not want to understand your contributions!" <- These are your words not mine.  I hard forked V3 for security reasons and ill Hard fork V4 for security reasons.  What do you want, i made it clear as day, i will no longer contribute anything to such an ignorant dev. Your ego is insuffearble and your code secuity practices are downright dangerous. Donald does fantastic work, while ive had to argue with you on everything. All just to keep a convoluted entrypoint script & dockerfile so you can say you wrote it, who cares. I just want a secure dashboard to mange wireguard in a container across x86/64, arm64 & arm32 cpus, even better if it can be done headless.

**Comment by DaanSelen** [comment]
Where did I say "Hi Noxcis, I do not want to understand your contributions!". 

Weird that you see asking for reasons as ignorant instead of trusting blind authority, which is even more dangerous.
I've asked to work with you, but you only offer the choice: "my way or the highway".

Unfortunate you react as a child in my opinion. Open-source should be with reasons and explanations instead of "your product bad, mine better".
